bee-common/bee-common/src/packable/mod.rs

```diff
@@ -0,0 +1,266 @@
+    /// Unpack this value from the given `Unpacker`.
+    fn unpack<U: Unpacker>(unpacker: &mut U) -> Result<Self, UnpackError<Self::Error, U::Error>>;
+    /// The size of the value in bytes after being packed.
+    fn packed_len(&self) -> usize;
+}
+
+macro_rules! impl_packable_for_int {
+    ($ty:ty) => {
+        impl Packable for $ty {
+            type Error = Infallible;
+
+            fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), P::Error> {
+                packer.pack_bytes(&self.to_le_bytes())
+            }
+
+            fn unpack<U: Unpacker>(unpacker: &mut U) -> Result<Self, UnpackError<Self::Error, U::Error>> {
+                let mut bytes = [0u8; core::mem::size_of::<Self>()];
+                unpacker.unpack_bytes(&mut bytes).map_err(UnpackError::Unpacker)?;
+                Ok(Self::from_le_bytes(bytes))
+            }
+
+            fn packed_len(&self) -> usize {
+                core::mem::size_of::<Self>()
+            }
+        }
+    };
+}
+
+impl_packable_for_int!(u8);
+impl_packable_for_int!(u16);
+impl_packable_for_int!(u32);
+impl_packable_for_int!(u64);
+#[cfg(has_u128)]
+impl_packable_for_int!(u128);
+
+/// `usize` integers are packed and unpacked as `u64` integers according to the spec.
+impl Packable for usize {
+    type Error = Infallible;
+
+    fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), P::Error> {
+        (*self as u64).pack(packer)
+    }
+
+    fn unpack<U: Unpacker>(unpacker: &mut U) -> Result<Self, UnpackError<Self::Error, U::Error>> {
+        Ok(unpacker.unpack_infallible::<u64>().map_err(UnpackError::Unpacker)? as usize)
+    }
+
+    fn packed_len(&self) -> usize {
+        core::mem::size_of::<u64>()
+    }
+}
+
+impl_packable_for_int!(i8);
+impl_packable_for_int!(i16);
+impl_packable_for_int!(i32);
+impl_packable_for_int!(i64);
+#[cfg(has_i128)]
+impl_packable_for_int!(i128);
+
+/// `isize` integers are packed and unpacked as `i64` integers according to the spec.
+impl Packable for isize {
+    type Error = Infallible;
+
+    fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), P::Error> {
+        (*self as i64).pack(packer)
+    }
+
+    fn unpack<U: Unpacker>(unpacker: &mut U) -> Result<Self, UnpackError<Self::Error, U::Error>> {
+        Ok(unpacker.unpack_infallible::<i64>().map_err(UnpackError::Unpacker)? as isize)
+    }
+
+    fn packed_len(&self) -> usize {
+        core::mem::size_of::<i64>()
+    }
+}
+
+impl Packable for bool {
+    type Error = Infallible;
+
+    /// Booleans are packed as `u8` integers following Rust's data layout.
+    fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), P::Error> {
+        (*self as u8).pack(packer)
+    }
+
+    /// Booleans are unpacked if the byte used to represent them is non-zero.
+    fn unpack<U: Unpacker>(unpacker: &mut U) -> Result<Self, UnpackError<Self::Error, U::Error>> {
+        Ok(unpacker.unpack_infallible::<u8>().map_err(UnpackError::Unpacker)? != 0)
```

Comment:
Shouldn't `unpack_infallible ` be a method of packable instead ?
Also, what's the point of calling `unpack_infallible` since we still have to map it to the same type as if we used unpack_bytes ?

Reply:
> Shouldn't `unpack_infallible ` be a method of packable instead ?
Yes but I couldn't figure out how to do that because it should only work for `Packable<Error = Infallible>` and I didn't want to add another trait just for that.
> Also, what's the point of calling `unpack_infallible` since we still have to map it to the same type as if we used unpack_bytes ?
That you don't need to map the `Packable::Error` error too because it doesn't happen. Otherwise you'd have to do something like
```rust
match unpack_error {
    UnpackError::Unpacker(err) => UnpackError::Unpacker(err),
    UnpackError::Packable(infallible) => match infallible {},
}
```
an alternative could be adding special methods for `UnpackError` if the packable error is `Infallible`.